File: Programação Dinâmica/lcs-top-down.cpp
/**
Autor: Daniel Saad Nogueira Nunes
Nome do Problema:
ID do Problema: 
url do Problema: 
Técnica:
Complexidade:
Comentários:
**/

#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;


int lcs_top_down(const string& x,const string& y, int n,int m,vector<vector<int>>& c){
	if(n==0 || m==0){
		return 0;
	}
	if(c[n-1][m-1]!=-1){
		return c[n-1][m-1];
	}
	if(x[n-1]==y[m-1]){
		c[n-1][m-1] = 1+ lcs_top_down(x,y,n-1,m-1,c);
		return (c[n-1][m-1]);
	}
	else{
		c[n-1][m-1] = max(lcs_top_down(x,y,n-1,m,c),lcs_top_down(x,y,n,m-1,c));
		return c[n-1][m-1];
	}
}

int lcs(const string& x,const string& y){
	vector<vector<int>> c(x.size(),vector<int>(y.size(),-1));
	return(lcs_top_down(x,y,x.size(),y.size(),c));
}

int main(){
	std::ios::sync_with_stdio(false);
	string x,y;
	cout << "Digite as sequencias X e Y: ";
	cin >> x >> y;
	cout << "A maior subsequencia comum de X e Y tem tamanho: " 
	<< lcs(x,y) << endl;
	return 0;
}

File: Programação Dinâmica/_lcs_top_down.java
package ProgramaçãoDinâmica;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Scanner;

public class _lcs_top_down {

	public static void main(String[] args) {
		Scanner entrada = new Scanner(System.in);
		String x,y;
		System.out.println("Digite as sequencias X e Y: ");
		x=entrada.nextLine();
		y=entrada.nextLine();
		System.out.println("A maior subsequencia comum de X e Y tem tamanho: " 
		+ lcs(x,y) );

	}

	private static Integer lcs(String x, String y) {
		List<List<Integer>> c =	new ArrayList<List<Integer>>(Collections.nCopies(x.length(),
				new ArrayList<Integer>(Collections.nCopies(y.length(),-1))));
		//vector<vector<int>> c(x.size(),     vector<int>(y.size(),-1));
		return(lcs_top_down(x,y,x.length(),y.length(),c));
	}

	private static Integer lcs_top_down(String x, String y,int n,int m, List<List<Integer>> c) {
		if(n==0 || m==0){
			return 0;
		}
		if(c.get(n-1).get(m-1)!=-1){
			return c.get(n-1).get(m-1);
		}
		if(x.charAt(n-1)==y.charAt(m-1)){
			c.get(n-1).set(m-1, 1+ lcs_top_down(x,y,n-1,m-1,c));
			return  c.get(n-1).get(m-1);
		}
		else{
			c.get(n-1).set(m-1, Math.max(lcs_top_down(x,y,n-1,m,c),lcs_top_down(x,y,n,m-1,c)));
			return c.get(n-1).get(m-1);
		}
	}

}
